### PR TITLE
Accept formatted JSON in the build request to specify the job config.

### PR DESCRIPTION
**Pull request by tjlee0909**
An example of what the JSON params look like when making a git-type build request:
```php
$params = [
        'type' => 'git',
        'url' => 'ssh://path.to.source.control/projectname',
        'branch' => 'master',
        'job_name' => 'CustomJobNameDoesntMatter',
        'config' => [
            'setup_build' => [
                'do some setup 1',
                'do some setup 2',
            ],
            'commands' => [
                'run some commands 1',
                'run some commands 2',
            ],
            'atomizers' => [
                [ 'TESTPATH' => 'find $PROJECT_DIR/ -name ...' ]
            ],
            'max_executors' => 100,
        ]
];
```

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @tjlee0909 has signed the CLA. Thanks for the pull request!

**Comment by tjlee0909** [comment]
An example of what the JSON params look like when making a git-type build request:

$params = [
        'type' => 'git',
        'url' => 'ssh://path.to.source.control/projectname',
        'branch' => 'master',
        'job_name' => 'CustomJobNameDoesntMatter',
        'config' => [
            'setup_build' => [
                'do some setup 1',
                'do some setup 2',
            ],
            'commands' => [
                'run some commands 1',
                'run some commands 2',
            ],
            'atomizers' => [
                [ 'TESTPATH' => 'find $PROJECT_DIR/ -name ...' ]
            ],
            'max_executors' => 100,
        ]
];

File: app/master/build_request.py [hunk 1]
```diff
@@ -22,6 +22,14 @@ class BuildRequest(object):
         "job_name": "clusterrunner_configured_job",
```
Comment (josephharrington) [inline app/master/build_request.py]: "job_name" becomes optional if "config" is specified, right?

Comment (tjlee0909) [inline app/master/build_request.py]: Functionally, yeah. But I was thinking about what would display in all of the dashboards we have, and was worried that they will just break. What do you think I should do?

Comment (josephharrington) [inline app/master/build_request.py]: Ah, true. I suppose ClusterRunner could just set the job_name value to some constant like "_adhoc_" if config was specified without a job name.

Comment (tjlee0909) [inline app/master/build_request.py]: I'm actually more worried about the 'job_name' field in the request param, more than the name field in the JobConfig, as that might be the value that some dashboards are utilizing.

That's something that should be fixed though, if it does that.

Comment (josephharrington) [inline app/master/build_request.py]: Ah yeah, that's what I meant. Maybe outside scope for this specific change then.

Comment (tjlee0909) [inline app/master/build_request.py]: That's correct, I'm trying to weasel myself out of more work.

File: test/functional/job_configs.py [hunk 2]
```diff
@@ -24,21 +24,16 @@ def __init__(self, config, expected_to_fail, expected_num_subjobs, expected_num_
...
+        'posix':
+            {
+                'commands': ['echo $TOKEN > $ARTIFACT_DIR/result.txt'],
+                'atomizers': [{'TOKEN': 'seq 0 4 | xargs -I {} echo "This is atom {}"'}]
+            },
+        'nt':
+            {
+                'commands': ['echo !TOKEN!> !ARTIFACT_DIR!\\result.txt'],
+                'atomizers': [{'TOKEN': 'FOR /l %n in (0,1,4) DO @echo This is atom %n'}]
+            },
```
Comment (josephharrington) [inline test/functional/job_configs.py]: I'd like to keep these functional tests as yaml strings if possible -- I think it's a good thing that the test data is in the raw format. Maybe we should change the framework instead?

Comment (tjlee0909) [inline test/functional/job_configs.py]: Sure, I can do that.

**Comment by josephharrington** [comment]
:+1:

**Comment by aptxkid** [comment]
:+1:

**Pull request merged**